File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool solve(vector<vector<char>>& board, string& word, int pointer, int i, int j, int n, int m, vector<vector<bool>>& visited,vector<vector<int>>& dir) {
        if (pointer == word.size()) return true;
        if (i < 0 || i >= n || j < 0 || j >= m || board[i][j] != word[pointer] || visited[i][j]) return false;
        visited[i][j] = true;
        for (auto d : dir) {
            if (solve(board, word, pointer + 1, i+d[0], j+d[1], n, m, visited,dir)) {
                return true;
            }
        }
        visited[i][j] = false;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        int n = board.size();
        int m = board[0].size();
                vector<vector<int>> dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        unordered_map<char,int> mp;
        for(auto r:board) {
            for(auto c:r) mp[c]++;
        }
        for(auto c:word) {
            mp[c]-=1;
            if(mp[c]<0) return false;
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (solve(board, word, 0, i, j, n, m, visited,dir)) return true;
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.py
from collections import defaultdict,deque
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        #m*n
        wordLength = len(word)

        #First count frequencies and see if its even possible or not
        freqCount = defaultdict(int)
        for row in board:
            for letter in row:
                freqCount[letter]+=1
        for letter in word:
            freqCount[letter]-=1
            if freqCount[letter]<0: return False
        
        visited = [[0 for _ in range(n)] for _ in range(m)]
        def dfs(i,j,pointer):
            if pointer==wordLength: return True
            if i<0 or j<0 or i>=m or j>=n: return False
            if visited[i][j]: return False
            if word[pointer]!=board[i][j]: return False
            if word[pointer]==board[i][j]:
                visited[i][j]=1

                result = (dfs(i+1,j,pointer+1) or dfs(i,j+1,pointer+1) or dfs(i-1,j,pointer+1) or dfs(i,j-1,pointer+1))
                visited[i][j]=0
                return result
           
        
        for i in range(m):
            for j in range(n):
                    if dfs(i,j,0)==True:
                        return True
        return False
            
        
    
        